micro_http/src/connection.rs

```diff
@@ -0,0 +1,634 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::collections::VecDeque;
+use std::io::{Read, Write};
+
+use common::ascii::{CR, CRLF_LEN, LF};
+use common::Body;
+pub use common::{ConnectionError, RequestError};
+use headers::Headers;
+use request::{find, Request, RequestLine};
+use response::{Response, StatusCode};
+
+/// Describes the state machine of an HTTP connection.
+pub enum ConnectionState {
+    WaitingForRequestLine,
+    WaitingForHeaders,
+    WaitingForBody,
+    RequestReady,
+}
+
+/// A wrapper over a HTTP Connection.
+pub struct HttpConnection<T> {
+    request: Option<Request>,
+    stream: T,
+    state: ConnectionState,
+    buffer: [u8; 1024],
+    start_index: usize,
+    body_vec: Vec<u8>,
+    body_bytes_to_be_read: i32,
+    parsed_requests: VecDeque<Request>,
+    response_queue: VecDeque<Response>,
+    response_buffer: Option<Vec<u8>>,
+}
+
+impl<T: Read + Write> HttpConnection<T> {
+    /// Creates a new default connection.
+    pub fn new(stream: T) -> Self {
+        HttpConnection {
+            request: None,
+            stream,
+            state: ConnectionState::WaitingForRequestLine,
+            buffer: [0; 1024],
```

Comment:
Using `Vec::with_capacity(1024)` doesn't guarantee that the vec will always have a length of 1024. If you push an element on an already full vec that was initialized using `with_capacity` the operation will not fail, it will just reallocate.

A vec instead of a slice would only add complexity to the code and no visible performance or readablility improvement as we would still have to use a slice for `read_exact` and also resize the vec every time. It is also heap allocated, as opposed to the current implementation which allocates on the stack and it prevents exceeding 1024 bytes.